internal/graph/check_fast_path.go

```diff
@@ -701,61 +661,38 @@ func (c *LocalChecker) breadthFirstRecursiveMatch(ctx context.Context, req *Reso
 
 		if err != nil {
 			concurrency.TrySendThroughChannel(ctx, checkOutcome{err: err}, checkOutcomeChan)
-			// TODO: TBD if we hard exit here, if yes, the channel needs to be closed
 			continue
 		}
-		c := make(chan *iterator.Msg, 1)
-		if !concurrency.TrySendThroughChannel(ctx, &iterator.Msg{Iter: mapper}, c) {
-			mapper.Stop() // will not be received to be cleaned up
-		}
-		close(c)
-		chans = append(chans, c)
-	}
-	leftChan := fanInIteratorChannels(ctx, chans, c.concurrencyLimit)
-	leftOpen := true
-	defer func() {
-		if leftOpen {
-			go drainIteratorChannel(leftChan)
-		}
-	}()
-	nextUsersetLevel := hashset.New()
-
-ConsumerLoop:
-	for {
-		select {
-		case <-ctx.Done():
-			close(checkOutcomeChan)
-			return
-		case msg, ok := <-leftChan:
-			if !ok {
-				// nothing was found in this level, break to proceed into the next one
-				leftOpen = false
-				break ConsumerLoop
-			}
-			for {
-				t, err := msg.Iter.Next(ctx)
-				if err != nil {
-					msg.Iter.Stop()
-					if storage.IterIsDoneOrCancelled(err) {
-						break
-					}
-					concurrency.TrySendThroughChannel(ctx, checkOutcome{err: err}, checkOutcomeChan)
-					close(checkOutcomeChan)
-					return
+		// if the pool is short-circuited, the iterator should be stopped
+		defer mapper.Stop()
+		pool.Go(func(ctx context.Context) error {
+			objectToUsersetMessageChan := streamedLookupUsersetFromIterator(ctx, mapper)
+			for usersetMsg := range objectToUsersetMessageChan {
```

Comment:
Do we need to care about context timeout?

Reply:
That should be handled by `streamedLookupUsersetFromIterator`.